code/modules/reagents/chemistry/machinery/smoke_machine.dm

```diff
@@ -0,0 +1,139 @@
+	var/smoke_test
+	smoke_test = locate(/obj/effect/particle_effect/smoke) in T.contents
+	if(on && (!smoke_test))
+		var/datum/effect_system/smoke_spread/chem/smoke_machine/smoke = new()
+		smoke.set_up(reagents, setting, efficiency, get_turf(src))
+		smoke.start()
+
+/obj/machinery/smoke_machine/attackby(obj/item/I, mob/user, params)
+	add_fingerprint(user)
+	if(istype(I, /obj/item/reagent_containers))
+		var/obj/item/reagent_containers/RC = I
+		var/units = RC.reagents.trans_to(src, RC.amount_per_transfer_from_this)
+		if(units)
+			to_chat(user, "<span class='notice'>You transfer [units] units of the solution to [src].</span>")
+			return
+	if(default_unfasten_wrench(user, I))
+		return
+	.=..()
+
+
+/obj/machinery/smoke_machine/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, \
+										datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "smoke_machine", name, 450, 350, master_ui, state)
+		ui.open()
+
+
+/obj/machinery/smoke_machine/ui_data(mob/user)
+	var/data = list()
+	var/TankContents[0]
+	var/TankCurrentVolume = 0
+	if(reagents.reagent_list.len)
+		for(var/datum/reagent/R in reagents.reagent_list)
+			TankContents.Add(list(list("name" = R.name, "volume" = R.volume))) // list in a list because Byond merges the first list...
+			TankCurrentVolume += R.volume
+	data["TankContents"] = TankContents
+	data["isTankLoaded"] = reagents.total_volume ? 1 : 0
+	data["TankCurrentVolume"] = reagents.total_volume ? reagents.total_volume : null
+	data["TankMaxVolume"] = reagents.maximum_volume
+	data["active"] = on
+	data["setting"] = setting
+	data["screen"] = screen
+	data["analyzeVars"] = analyzeVars
+	return data
+
+/obj/machinery/smoke_machine/ui_act(action, params)
+	if(..() || !anchored)
+		return
+	switch(action)
+		if("purge")
+			reagents.clear_reagents()
+			. = TRUE
+
+		if("analyze")
+			var/datum/reagent/R = GLOB.chemical_reagents_list[params["id"]]
+			if(R)
+				var/state = "Unknown"
+				if(initial(R.reagent_state) == 1)
+					state = "Solid"
+				else if(initial(R.reagent_state) == 2)
+					state = "Liquid"
+				else if(initial(R.reagent_state) == 3)
+					state = "Gas"
+				var/const/P = 3 //The number of seconds between life ticks
+				var/T = initial(R.metabolization_rate) * (60 / P)
+				analyzeVars = list("name" = initial(R.name), "state" = state, "color" = initial(R.color), "description" = initial(R.description), "metaRate" = T, "overD" = initial(R.overdose_threshold), "addicD" = initial(R.addiction_threshold))
+				screen = "analyze"
+		if("setting")
+			var/amount = text2num(params["amount"])
+			if (locate(amount) in possible_settings)
+				setting = amount
+				. = TRUE
+		if("power")
+			on = !on
+			if(on)
+				log_admin("[key_name(usr)] activated a smoke machine that contains the following at [COORD(src)]:")
+				for(var/datum/reagent/R in reagents.reagent_list)
```

Comment:
add_logs that it was turned on

Reply:
message_admins as well

Reply:
I really don't think admins want to get spammed every time this thing is turned on.